clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1107,25 +1122,33 @@ public void assign(Collection<TopicPartition> partitions) {
             if (this.subscriptions.hasNoSubscriptionOrUserAssignment())
                 throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
 
-            // poll for new data until the timeout expires
-            long start = time.milliseconds();
             long remaining = timeout;
             do {
-                Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollOnce(remaining);
+
+                client.maybeTriggerWakeup();
+
+                while (!internalUpdateAssignmentMetadataIfNeeded(A_MONTH)) {
```

Comment:
Shouldn't the time spent in this loop be counted toward ```remaining``` ?

Reply:
Good eye. But this is the "old" method, whose behavior I'm preserving. So this is a re-expression of the existing behavior, that metadata update will block arbitrarily long and not count against the timeout, followed by the fetch, to which the timeout applies.